src/coreclr/debug/runtimeinfo/datadescriptor.h

```diff
@@ -0,0 +1,139 @@
+//   size k (bytes - usually sizeof(SomeNativeType)) or specify that the type's size is not provided
+//   It is important that CDAC_TYPE_SIZE or CDAC_TYPE_INDETERMINATE immediately follows
+//   CDAC_TYPE_BEGIN
+//
+// CDAC_TYPE_FIELD(cdacTypeIdentifier, cdacFieldTypeIdentifier, cdacFieldName, k) specifies the
+//   field of "cdacTypeIdentifier" that has name cdacFieldName and has the type
+//   "cdacFieldtypeIdentifier" located at offset k in the type layout.  k is usually
+//   offsetof(SomeClass, m_FieldName) if the field is public
+//
+//     if the field is private, the convention is that SomeClass declares a friend struct
+//     cdac_offsets<T> and provides a specialization of cdac_offsets<T> with a public constexpr
+//     size_t member that holds the offset:
+//
+//     class MyClass {
+//       private:
+//         void* m_myField;
+//         friend template<typename T> cdac_offsets<T>;
+//     };
+//     template<> struct cdac_offsets<MyClass> {
+//       static constexpr size_t MyField = offsetof(MyClass, m_myField);
+//     };
+//
+//     then the field layout can be specified as
+//     CDAC_TYPE_FIELD(MyClassLayout, pointer, MyField, cdac_offsets<MyClass>::MyField)
+//  There can be zero or more CDAC_TYPE_FIELD entries per type layout
+//
+// CDAC_TYPE_END(cdacTypeIdentifier)  specifies the end of the type layout for cdacTypeIdentifier
+//
+// In <globals> the format is:
+//
+// CDAC_GLOBAL(cdacGlobalName, cdacTypeIdentifier, value)
+// or
+// CDAC_GLOBAL_POINTER(cdacGlobalName, cdacTypeIdentifier, address)
+//
+// Zero or more globals can be defined
+//
+// if a global is given with CDAC_GLOBAL(), `value` should be a constexpr uint64_t (or convertible
+// to uint64_t) for example, it can be a literal constant or a preprocessor definition
+//
+// if a global is a CDAC_GLOBAL_POINTER(), address should be a constexpr pointer or a constexpr
+// uintptr_t
+//
+//
+//
+// This file is compiled using the target architecture.  Preprocessor defines for the target
+// platform will be available.  It is ok to use `#ifdef`.
+
+#ifndef CDAC_BASELINE
+#define CDAC_BASELINE(identifier)
+#endif
+#ifndef CDAC_TYPES_BEGIN
+#define CDAC_TYPES_BEGIN()
+#endif
+#ifndef CDAC_TYPE_BEGIN
+#define CDAC_TYPE_BEGIN(tyname)
+#endif
+#ifndef CDAC_TYPE_SIZE
+#define CDAC_TYPE_SIZE(k)
+#endif
+#ifndef CDAC_TYPE_INDETERMINATE
+#define CDAC_TYPE_INDETERMINATE(tyname)
+#endif
+#ifndef CDAC_TYPE_FIELD
+#define CDAC_TYPE_FIELD(tyname,fieldtyname,fieldname,off)
+#endif
+#ifndef CDAC_TYPE_END
+#define CDAC_TYPE_END(tyname)
+#endif
+#ifndef CDAC_TYPES_END
+#define CDAC_TYPES_END()
+#endif
+#ifndef CDAC_GLOBALS_BEGIN
+#define CDAC_GLOBALS_BEGIN()
+#endif
+#ifndef CDAC_GLOBAL
+#define CDAC_GLOBAL(globalname,tyname,val)
+#endif
+#ifndef CDAC_GLOBAL_POINTER
+#define CDAC_GLOBAL_POINTER(globalname,addr)
+#endif
+#ifndef CDAC_GLOBALS_END
+#define CDAC_GLOBALS_END()
+#endif
+
+CDAC_BASELINE("empty")
```

Comment:
It's hard to have completely optional data with the preprocessor since everything here expands to a single global initialization expression. I would need some keyword for "no baseline".

Reply:
Follow-up work: https://github.com/dotnet/runtime/issues/101257